In [ ]:
#Import Libraries
import networkx as nx
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [38]:
#load CSV Files into dataframe
csv_file='SFlow Data -Sem 1 2020.csv'
df = pd.read_csv(csv_file, header=None)

In [39]:
#Check if there are any column that is NaN
df.isnull().head(5)

0      1      2      3      4      5      6      7      8      9   \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   
2  False  False  False  False  False  False  False  False  False  False   
3  False  False  False  False  False  False  False  False  False  False   
4  False  False  False  False  False  False  False  False  False  False   

      10     11     12     13     14     15     16     17     18     19  
0  False  False  False  False  False  False  False  False  False  False  
1  False  False  False  False  False  False  False  False  False  False  
2  False  False  False  False  False  False  False  False  False  False  
3  False  False  False  False  False  False  False  False  False  False  
4  False  False  False  False  False  False  False  False  False  False

In [40]:
#Insert Header according to lab sheet
df.columns = ["FLOW",
              "sflow_agent_address",
              "inputPort",
              "outputPort",
              "src_MAC",
              "dst_MAC",
              "ethernet_type",
              "in_vlan",
              "out_vlan",
              "src_IP",
              "dst_IP",
              "IP_protocol",
              "ip_tos",
              "ip_ttl",
              "src_port_addr",
              "dst_port_addr",
              "tcp_flags",
              "packet_size",
              "IP_size",
              "sampling_rate",
             ]

In [41]:
df.head(5)

FLOW sflow_agent_address  inputPort  outputPort       src_MAC  \
0  FLOW       203.30.38.251        129         193  00135f21bc80   
1  FLOW       203.30.38.251        129         193  00135f21bc80   
2  FLOW       203.30.38.251        131         193  001cb0c88e40   
3  FLOW       203.30.38.251        135         129  002688cd5fc7   
4  FLOW       203.30.38.251        193         130  0031466b23cf   

        dst_MAC ethernet_type  in_vlan  out_vlan           src_IP  \
0  0031466b23cf        0x0800     11.0       919    155.69.160.77   
1  0031466b23cf        0x0800     11.0       919    155.69.160.78   
2  0031466b23cf        0x0800     43.0       919  192.122.131.134   
3  00135f21bc80        0x0800    919.0        11   54.169.229.179   
4  00239cd087c1        0x0800    919.0       919   209.85.229.247   

           dst_IP  IP_protocol ip_tos  ip_ttl  src_port_addr  dst_port_addr  \
0  74.125.130.141            6   0x60      59        57577.0            443   
1  74.125.203.128            6   0x00      59        34809.0            443   
2    74.125.10.12            6   0x00      57        50549.0            443   
3  155.69.191.254           17   0x00      57        16285.0          39641   
4  137.132.228.34           17   0x00      63          443.0          32817   

  tcp_flags  packet_size  IP_size  sampling_rate  
0      0x10           74       52           2048  
1      0x10         1442     1420           2048  
2      0x10          102       80           2048  
3      0x10          580      562           2048  
4      0x00         1396     1378           2048

In [42]:
#Check if column "FLOW" just contain FLOW
df["FLOW"].unique()

array(['FLOW', 'CNTR'], dtype=object)

In [43]:
# #Extract rows that is FLOW
flow_df = df[df["FLOW"] == "FLOW"]

In [44]:
df

FLOW sflow_agent_address  inputPort  outputPort       src_MAC  \
0       FLOW       203.30.38.251        129         193  00135f21bc80   
1       FLOW       203.30.38.251        129         193  00135f21bc80   
2       FLOW       203.30.38.251        131         193  001cb0c88e40   
3       FLOW       203.30.38.251        135         129  002688cd5fc7   
4       FLOW       203.30.38.251        193         130  0031466b23cf   
...      ...                 ...        ...         ...           ...   
203998  FLOW       203.30.38.251        129         193  00135f21bc80   
203999  FLOW       203.30.38.251        193         130  0031466b23cf   
204000  FLOW       203.30.38.251        200         130  80711fc76001   
204001  FLOW       203.30.38.251        193         130  0031466b23cf   
204002  FLOW       203.30.38.251        193         130  0031466b23cf   

             dst_MAC ethernet_type  in_vlan  out_vlan           src_IP  \
0       0031466b23cf        0x0800     11.0       919    155.69.160.77   
1       0031466b23cf        0x0800     11.0       919    155.69.160.78   
2       0031466b23cf        0x0800     43.0       919  192.122.131.134   
3       00135f21bc80        0x0800    919.0        11   54.169.229.179   
4       00239cd087c1        0x0800    919.0       919   209.85.229.247   
...              ...           ...      ...       ...              ...   
203998  0031466b23cf        0x0800     11.0       919   155.69.203.215   
203999  00239cd087c1        0x0800    919.0       919   173.194.22.204   
204000  00239cd087c1        0x0800    280.0       919  169.229.205.229   
204001  00239cd087c1        0x0800    919.0       919     74.125.10.40   
204002  00239cd087c1        0x0800    919.0       919   173.194.22.204   

                dst_IP  IP_protocol ip_tos  ip_ttl  src_port_addr  \
0       74.125.130.141            6   0x60      59        57577.0   
1       74.125.203.128            6   0x00      59        34809.0   
2         74.125.10.12            6   0x00      57        50549.0   
3       155.69.191.254           17   0x00      57        16285.0   
4       137.132.228.34           17   0x00      63          443.0   
...                ...          ...    ...     ...            ...   
203998     74.125.10.7            6   0x00     122        55177.0   
203999   137.132.17.41           17   0x00      63          443.0   
204000  137.132.22.254           17   0x00     247          443.0   
204001  137.132.188.86            6   0x00      63          443.0   
204002   137.132.17.41           17   0x00      63          443.0   

        dst_port_addr tcp_flags  packet_size  IP_size  sampling_rate  
0                 443      0x10           74       52           2048  
1                 443      0x10         1442     1420           2048  
2                 443      0x10          102       80           2048  
3               39641      0x10          580      562           2048  
4               32817      0x00         1396     1378           2048  
...               ...       ...          ...      ...            ...  
203998            443      0x10           62       40           2048  
203999          54938      0x00         1396     1378           2048  
204000           3090      0x00         1487     1465           2048  
204001          54901      0x10         1438     1420           2048  
204002          54938      0x10         1396     1378           2048  

[204003 rows x 20 columns]

# Exercise 3A: TOP TALKERS AND LISTENERS

### Top 5 Talker & Packet send

In [45]:
##Get the Top 5 Talker and store into dict
toptalker_count = flow_df["src_IP"].value_counts().head(5).to_dict()

In [46]:
print(f"Top 5 Talker   : Number of times sending")
for key, item in toptalker_count.items():
    print(f"{key:<15}: {item}")

Top 5 Talker   : Number of times sending
103.26.47.233  : 9646
13.107.4.50    : 4950
155.69.160.78  : 4563
130.14.250.7   : 3914
173.194.22.215 : 2896


In [47]:
#Also store the packets that the Top 5 Talker send
toptalker_pkt = toptalker_count
for key in toptalker_pkt:
    toptalker_pkt[key] = flow_df.loc[flow_df["src_IP"] == key]["packet_size"].sum()

In [48]:
print(f"Top 5 Talker   : Total packet size with MAC header")
for key, item in toptalker_pkt.items():
    print(f"{key:<15}: {item}")

Top 5 Talker   : Total packet size with MAC header
103.26.47.233  : 14232914
13.107.4.50    : 7327262
155.69.160.78  : 2524871
130.14.250.7   : 5957108
173.194.22.215 : 4073571


### Top 5 Listener & Packet received

In [49]:
##Get the Top 5 Listener and store into dict
toplistener_count = flow_df["dst_IP"].value_counts().head(5).to_dict()

In [50]:
print(f"Top 5 Listener: Number of times receiving")
for key, item in toplistener_count.items():
    print(f"{key:<14}: {item}")

Top 5 Listener: Number of times receiving
103.22.221.73 : 9646
137.132.228.33: 7835
137.132.228.29: 5964
137.132.228.42: 4987
103.37.198.100: 3915


In [51]:
#Also store the packets that the Top 5 Listener received
toplistener_pkt = toplistener_count
for key in toplistener_pkt:
    toplistener_pkt[key] = flow_df.loc[flow_df["dst_IP"] == key]["packet_size"].sum()

In [52]:
print(f"Top 5 Listener: Total packet size with MAC header")
for key, item in toplistener_pkt.items():
    print(f"{key:<14}: {item}")

Top 5 Listener: Total packet size with MAC header
103.22.221.73 : 14232914
137.132.228.33: 10900079
137.132.228.29: 8149270
137.132.228.42: 6826231
103.37.198.100: 5957182


# Exercise 3B: TRANSPORT PROTOCOL

In [53]:
#TCP - 6, UDP - 17
totalTCP = (flow_df["IP_protocol"] == 6).sum()
totalUDP = (flow_df["IP_protocol"] == 17).sum()

print(f"Number of TCP protocol used: {totalTCP}")
print(f"Number of UDP protocol used: {totalUDP}")
print(f"Percentage of TCP protocol used: {totalTCP/len(flow_df)*100:.2f}%")
print(f"Percentage of UDP protocol used: {totalUDP/len(flow_df)*100:.2f}%")

Number of TCP protocol used: 155799
Number of UDP protocol used: 45377
Percentage of TCP protocol used: 76.83%
Percentage of UDP protocol used: 22.38%


### Percentage for all Protocol used

In [54]:
packet_ip = []

ip_proto_df = pd.DataFrame(flow_df["IP_protocol"].value_counts())

for x in ip_proto_df.index:
    packet_ip.append(flow_df[flow_df["IP_protocol"] == x]["packet_size"].sum())

ip_proto_df["Percentage"] = (flow_df["IP_protocol"].value_counts()/len(flow_df)*100).round(decimals=2)
ip_proto_df.rename(columns={"IP_protocol": "# of Packets"},inplace=True)
ip_proto_df.insert(2, "Total Packets Size", packet_ip)
ip_proto_df.rename_axis('IP_protocol', axis=1)

IP_protocol  # of Packets  Percentage  Total Packets Size
6                  155799       76.83           152859000
17                  45377       22.38            49341573
47                    891        0.44              722079
50                    643        0.32              339994
1                      62        0.03                7349
58                      7        0.00                 738
41                      3        0.00                 318
0                       3        0.00                 204
2                       2        0.00                 108
89                      1        0.00                  94

# Exercise 3C: APPLICATION PROTOCOL

In [55]:
app_proto_df = pd.DataFrame(flow_df["dst_port_addr"].value_counts().head(5))

packet_app = []
for x in app_proto_df.index:
    packet_app.append(flow_df[flow_df["dst_port_addr"] == x]["packet_size"].sum())
    
app_proto_df.insert(1, "Total Packets Size", packet_app)
app_proto_df.rename(columns={"dst_port_addr": "# of Packets"},inplace=True)
app_proto_df.rename_axis("Port #", axis=1)

Port #  # of Packets  Total Packets Size
443            42975            18283059
80             11960             1258414
56800           3918             5960828
15000           2697             3864128
44678           1158             1706516

Port 443 is the most frequenly used application protocol. It means most traffic passed on HTTPS.

# Exercise 3D: TRAFFIC INTENSITY

In [56]:
#Assuming each packet is a bytes
#Assuming that SFlow traffic dataset was recorded for 15 sec.
#Packet Traffic
#Assuming sampling rate is 1/1000. Dataset is 1 sample. Hence, we multiply by a thousand
packet_traffic = flow_df["packet_size"].sum()*1000

#Convert to MB
packet_traffic_MB = packet_traffic/2**20
print(f'Packet Traffic (15s): {packet_traffic_MB:.2f}MB')

#IP Traffic
#Assuming sampling rate is 1/1000. Dataset is 1 sample. Hence, we multiply by a thousand
ip_traffic = flow_df["IP_size"].sum()*1000 

#Convert to MB
ip_traffic_MB = ip_traffic/2**20
print(f'IP Traffic (15s): {ip_traffic_MB:.2f}MB')

MAC_header = 1-ip_traffic_MB/packet_traffic_MB
print(f'Total MAC header in Packet Traffic (15s): {MAC_header*100:.2f}%')

Packet Traffic (15s): 193854.77MB
IP Traffic (15s): 189937.24MB
Total MAC header in Packet Traffic (15s): 2.02%


# OPTIONAL

### Most Popular Node provide Service on Port 80, Port 443

In [57]:
p80_df = pd.DataFrame(flow_df[["dst_MAC","dst_port_addr"]])

In [58]:
dups = p80_df.pivot_table(index = ["dst_MAC","dst_port_addr"], aggfunc ='size')
dups.sort_values(ascending=False).head(5)

dst_MAC       dst_port_addr
0031466b23cf  443              26933
5c5eabb2b0ce  443               5111
50878977e470  56800             3914
2c2172f7f437  80                3901
8cb64fe6b643  443               3282
dtype: int64

### Top 5 Communication Pair

In [61]:
ip_pair_df = pd.DataFrame(flow_df.groupby(['src_IP','dst_IP']).count()["FLOW"])
ip_pair_df = ip_pair_df.sort_values(by='FLOW',ascending=False)
ip_pair_df.columns=["Count"]
ip_pair_df.head(5)

Count
src_IP         dst_IP               
103.26.47.233  103.22.221.73    9646
130.14.250.7   103.37.198.100   3914
129.99.230.54  137.132.22.74    2689
104.44.201.147 202.21.159.244   2454
202.21.159.244 104.44.201.147   2154

Top 2 Communication Pair can be double checked with Top5Talker and Top5Listener

In [62]:
v_df = flow_df[["src_IP","src_port_addr","dst_IP","dst_port_addr"]]
v_df["src_port_addr"] = v_df["src_port_addr"].astype(int).astype(str)
v_df["dst_port_addr"] = v_df["dst_port_addr"].astype(str)
v_df = v_df.values.tolist()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [63]:
for x in v_df:
    x[0:2] = [':'.join(x[0:2])]
    x[1:3] = [':'.join(x[1:3])]
v_df = [tuple(x) for x in v_df]

In [64]:
#list for communication between hosts for graph with IP:port num
v_df

[('155.69.160.77:57577', '74.125.130.141:443'),
 ('155.69.160.78:34809', '74.125.203.128:443'),
 ('192.122.131.134:50549', '74.125.10.12:443'),
 ('54.169.229.179:16285', '155.69.191.254:39641'),
 ('209.85.229.247:443', '137.132.228.34:32817'),
 ('173.194.22.219:443', '137.132.228.29:50795'),
 ('172.217.26.197:443', '123.136.64.7:3815'),
 ('130.14.250.7:50189', '103.37.198.100:56800'),
 ('192.170.233.204:6885', '202.6.241.101:37213'),
 ('207.241.227.110:80', '192.248.8.114:40454'),
 ('103.26.47.233:36296', '103.22.221.73:40008'),
 ('74.125.10.59:443', '137.132.68.126:57765'),
 ('155.69.160.78:34809', '74.125.203.128:443'),
 ('155.69.110.100:64384', '104.66.3.203:443'),
 ('155.69.160.78:12327', '74.125.130.95:443'),
 ('137.132.228.29:4764', '152.61.133.67:443'),
 ('130.14.250.24:443', '123.136.64.31:42258'),
 ('209.85.229.252:443', '137.132.228.33:55556'),
 ('74.125.10.25:443', '137.132.228.33:53932'),
 ('103.26.47.233:36296', '103.22.221.73:40008'),
 ('155.69.221.177:51129', '58.27.86.3

In [ ]:
#Building of Graph with IP pair
H = nx.Graph()
for pair in v_df:
    H.add_edge(pair[0], pair[1])
    
plt.figure(figsize=(100,100)) 
nx.draw(H, with_labels = True, node_color='black')
plt.savefig("ip_comms_1.png") # save as png
plt.show()

### Visualizing Communication between IP Hosts

In [ ]:
#Reusing IP Pair did on top 5 communication pair
ip_pair_list = list(ip_pair_df.index)

#Building of Graph with IP pair
G = nx.Graph()
for pair in ip_pair_list:
    G.add_edge(pair[0], pair[1])
    
plt.figure(figsize=(100,100)) 
nx.draw(G, with_labels = True, node_color='black')
plt.savefig("ip_comms.png") # save as png
plt.show()